actioncable/lib/action_cable/connection/base.rb

```diff
@@ -0,0 +1,219 @@
+        @subscriptions  = ActionCable::Connection::Subscriptions.new(self)
+        @message_buffer = ActionCable::Connection::MessageBuffer.new(self)
+
+        @started_at = Time.now
+      end
+
+      # Called by the server when a new WebSocket connection is established. This configures the callbacks intended for overwriting by the user.
+      # This method should not be called directly. Rely on the #connect (and #disconnect) callback instead.
+      def process
+        logger.info started_request_message
+
+        if websocket.possible? && allow_request_origin?
+          websocket.on(:open)    { |event| send_async :on_open   }
+          websocket.on(:message) { |event| on_message event.data }
+          websocket.on(:close)   { |event| send_async :on_close  }
+
+          respond_to_successful_request
+        else
+          respond_to_invalid_request
+        end
+      end
+
+      # Data received over the cable is handled by this method. It's expected that everything inbound is JSON encoded.
+      # The data is routed to the proper channel that the connection has subscribed to.
+      def receive(data_in_json)
+        if websocket.alive?
+          subscriptions.execute_command ActiveSupport::JSON.decode(data_in_json)
+        else
+          logger.error "Received data without a live WebSocket (#{data_in_json.inspect})"
+        end
+      end
+
+      # Send raw data straight back down the WebSocket. This is not intended to be called directly. Use the #transmit available on the
+      # Channel instead, as that'll automatically address the correct subscriber and wrap the message in JSON.
+      def transmit(data)
+        websocket.transmit data
+      end
+
+      # Close the WebSocket connection.
+      def close
+        websocket.close
+      end
+
+      # Invoke a method on the connection asynchronously through the pool of thread workers.
+      def send_async(method, *arguments)
+        worker_pool.async.invoke(self, method, *arguments)
+      end
+
+      # Return a basic hash of statistics for the connection keyed with `identifier`, `started_at`, and `subscriptions`.
+      # This can be returned by a health check against the connection.
+      def statistics
+        {
+          identifier: connection_identifier,
+          started_at: @started_at,
+          subscriptions: subscriptions.identifiers,
+          request_id: @env['action_dispatch.request_id']
+        }
+      end
+
+      def beat
+        transmit ActiveSupport::JSON.encode(identifier: ActionCable::INTERNAL[:identifiers][:ping], message: Time.now.to_i)
+      end
+
+
+      protected
+        # The request that initiated the WebSocket connection is available here. This gives access to the environment, cookies, etc.
+        def request
+          @request ||= begin
+            environment = Rails.application.env_config.merge(env) if defined?(Rails.application) && Rails.application
```

Comment:
This is a smell. We should not access the application in the frameworks. We usually invert these dependencies adding a railtie.